rust/datafusion/src/physical_plan/hash_join.rs

```diff
@@ -0,0 +1,507 @@
+// E.g. [1, 2] -> [(0, 3), (1, 6), (0, 8)] indicates that (column1, column2) = [1, 2] is true
+// for rows 3 and 8 from batch 0 and row 6 from batch 1.
+type JoinHashMap = HashMap<Vec<GroupByScalar>, Vec<Index>>;
+type JoinLeftData = (JoinHashMap, Vec<RecordBatch>);
+
+/// join execution plan executes partitions in parallel and combines them into a set of
+/// partitions.
+#[derive(Debug)]
+pub struct HashJoinExec {
+    /// left side
+    left: Arc<dyn ExecutionPlan>,
+    /// right side
+    right: Arc<dyn ExecutionPlan>,
+    /// Set of common columns used to join on
+    on: HashSet<String>,
+    /// How the join is performed
+    join_type: JoinType,
+    /// The schema once the join is applied
+    schema: SchemaRef,
+}
+
+impl HashJoinExec {
+    /// Tries to create a new [HashJoinExec].
+    /// # Error
+    /// This function errors when it is not possible to join the left and right sides on keys `on`.
+    pub fn try_new(
+        left: Arc<dyn ExecutionPlan>,
+        right: Arc<dyn ExecutionPlan>,
+        on: &HashSet<String>,
+        join_type: &JoinType,
+    ) -> Result<Self> {
+        let left_schema = left.schema();
+        let right_schema = right.schema();
+        check_join_is_valid(&left_schema, &right_schema, &on)?;
+
+        let on = on.iter().map(|s| s.clone()).collect::<HashSet<_>>();
+
+        let schema = Arc::new(build_join_schema(
+            &left_schema,
+            &right_schema,
+            &on,
+            &join_type,
+        ));
+
+        Ok(HashJoinExec {
+            left,
+            right,
+            on: on.clone(),
+            join_type: join_type.clone(),
+            schema,
+        })
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for HashJoinExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.left.clone(), self.right.clone()]
+    }
+
+    fn with_new_children(
+        &self,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        match children.len() {
+            2 => Ok(Arc::new(HashJoinExec::try_new(
+                children[0].clone(),
+                children[1].clone(),
+                &self.on,
+                &self.join_type,
+            )?)),
+            _ => Err(DataFusionError::Internal(
+                "HashJoinExec wrong number of children".to_string(),
+            )),
+        }
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.right.output_partitioning()
+    }
+
+    async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        // merge all parts into a single stream
+        // this is currently expensive as we re-compute this for every part from the right
+        // TODO: Fix this issue: we can't share this state across parts on the right.
```

Comment:
When I tried to implement this, my approach was to store the materialized left-hand side in an `Arc<Mutex<_>>` and my plan was to use logic like this:

```rust
if (partition == 0) {
  // materialize left input and store in Arc<Mutex<_>>
} else {
 // wait for left input to be materialized
}
```

Reply:
I woulld be fine with us doing this as a follow up and get the expensive version of this functionally correct before we optimize too much.

Reply:
I agree with this suggestion

Reply:
I will investigate this `execute` limitation. I have an idea about it: I think that we should change `async execute` to return a vector of streams (one stream per part). I.e. instead of 

```rust
let stream_0 = node.execute(0)?.await;
let stream_1 = node.execute(1)?.await;
...
```

we use 

```rust
let streams = node.execute()?.await;
let stream_0 = streams[0];
let stream_1 = streams[1];
...
```

by doing this, we can make `node.execute()`'s implementation to compute the left side, and have that shared across all streams (`HashJoinStream`) without mutating the node (which IMO is good that it is immutable).

IMO we may not need an `Arc<Mutex<`, as that data is immutable: an `Arc` should be enough to share this across the streams (but again, I need to check this).